File: Trees/001 Binary Search Tree/Lowest Common Ancestor of a Binary Search Tree/IterativeSolution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        
        //root will not be null, as we will get result before that
        while(true) {
            int curr = root.val;
            if(curr>p.val && curr>q.val)
                //both of them lie on left
                root=root.left;
            else if(curr<p.val && curr<q.val)
                //both of them lie on right
                root=root.right;
            else
                //first time they split
                //or p/q = curr
                return root;
        }

    }
}

File: Trees/001 Binary Search Tree/Lowest Common Ancestor of a Binary Search Tree/IterativeSolution.kt
/**
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int = 0) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */

class Solution {
    fun lowestCommonAncestor(root: TreeNode?, p: TreeNode?, q: TreeNode?): TreeNode? {
        //root will not be null, as we will get result before that
        var rootC = root
        var curr = 0
        while(true) {
            curr = rootC!!.`val`
            if(curr>p!!.`val` && curr>q!!.`val`)
                //both of them lie on left
                rootC=rootC!!.left
            else if(curr<p!!.`val` && curr<q!!.`val`)
                //both of them lie on right
                rootC=rootC!!.right
            else
                //first time they split
                //or p/q = curr
                return rootC
        }
    }
}
